### PR TITLE
Reject POST and PUT requests with no Content-Length or Transfer-Encoding (#1130)

### PR DESCRIPTION
**Pull request by cesarbs**
#1130 

@halter73 @davidfowl @Tratcher

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/MessageBody.cs [hunk 1]
```diff
@@ -280,6 +281,13 @@ protected virtual void OnConsumedBytes(int count)
                 }
             }
 
+            // If we got here, request contains no Content-Length or Transfer-Encoding header.
+            // Reject with 411 Length Required.
+            if (HttpMethods.IsPost(context.Method) || HttpMethods.IsPut(context.Method))
```
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/MessageBody.cs]: No user code has run at this point. Do a strict reference equals.

File: test/Microsoft.AspNetCore.Server.KestrelTests/MessageBodyTests.cs [hunk 2]
```diff
@@ -196,13 +196,29 @@ public void ForThrowsWhenFinalTransferCodingIsNotChunked()
             using (var input = new TestInput())
             {
                 var ex = Assert.Throws<BadHttpRequestException>(() =>
-                    MessageBody.For(HttpVersion.Http10, new FrameRequestHeaders { HeaderTransferEncoding = "chunked, not-chunked" }, input.FrameContext));
+                    MessageBody.For(HttpVersion.Http11, new FrameRequestHeaders { HeaderTransferEncoding = "chunked, not-chunked" }, input.FrameContext));
```
Comment (halter73) [inline test/Microsoft.AspNetCore.Server.KestrelTests/MessageBodyTests.cs]: Why did this change?

File: test/Microsoft.AspNetCore.Server.KestrelTests/MessageBodyTests.cs [hunk 3]
```diff
@@ -196,13 +196,29 @@ public void ForThrowsWhenFinalTransferCodingIsNotChunked()
...
+        [Theory]
+        [InlineData("POST")]
+        [InlineData("PUT")]
+        public void ForThrowsWhenMethodRequiresLengthButNoContentLengthOrTransferEncodingIsSet(string method)
+        {
+            using (var input = new TestInput())
+            {
+                input.FrameContext.Method = method;
+                var ex = Assert.Throws<BadHttpRequestException>(() =>
+                    MessageBody.For(HttpVersion.Http11, new FrameRequestHeaders(), input.FrameContext));
```
Comment (halter73) [inline test/Microsoft.AspNetCore.Server.KestrelTests/MessageBodyTests.cs]: Should we test this for HTTP/1.0 too?

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/MessageBody.cs (hunk 1 above)
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/MessageBody.cs]: Were there any other methods you considered requiring a content length for?

File: test/Microsoft.AspNetCore.Server.KestrelTests/MessageBodyTests.cs (hunk 3 above)
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/MessageBodyTests.cs]:
```
If a request
contains an entity body and Content-Length is not specified, and the
server does not recognize or cannot calculate the length from other
fields, then the server should send a 400 (bad request) response.
```

https://tools.ietf.org/html/rfc1945#section-7.2.2

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/MessageBody.cs (hunk 1 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/MessageBody.cs]: Nope. IIS and nginx seem to require a length only for POST and PUT, so I'm doing the same here.

**Comment by cesarbs** [comment]
@halter73 Updated with early check for != GET.

**Comment by halter73** [comment]
Both travis builds are failing with "Block being garbage collected instead of returned to pool: IncomingStart, /Users/travis/build/aspnet/KestrelHttpServer/src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs, 66"

https://travis-ci.org/aspnet/KestrelHttpServer/builds/173249429

File: test/Microsoft.AspNetCore.Server.KestrelTests/MessageBodyTests.cs [hunk 4]
```diff
@@ -196,13 +196,45 @@ public void ForThrowsWhenFinalTransferCodingIsNotChunked()
...
+
+                Assert.Equal(411, ex.StatusCode);
+                Assert.Equal($"{method} request contains no Content-Length or Transfer-Encoding header", ex.Message);
+            }
+        }
+
+        [Theory]
+        [InlineData("POST")]
+        [InlineData("PUT")]
+        public void ForThrowsWhenMethodRequiresLengthButNoContentLengthSetHttp10(string method)
```
Comment (halter73) [inline test/Microsoft.AspNetCore.Server.KestrelTests/MessageBodyTests.cs]: My guess is that this test is throwing causing the block to be leaked.

Comment (halter73) [inline test/Microsoft.AspNetCore.Server.KestrelTests/MessageBodyTests.cs]: Never mind this issue repros in the dev branch. I found your issue https://github.com/aspnet/KestrelHttpServer/issues/1195.

**Comment by cesarbs** [comment]
Opened https://github.com/dotnet/corefx/issues/13369 to track test failures caused by hang on 0-byte async write to `SslStream`.

**Comment by cesarbs** [comment]
Rebased with b3aca0413c20226694502deb8649abf4e909ce0f.

**Pull request merged**